### ISSUE TITLE
SyncContext problem

### ISSUE THREAD

**Issue by MathieuGrdr** [beto-rodriguez/LiveCharts2#785]
I'm making a WPF app with MVVM architecture and Livecharts2, and I'm updating a chart every 500ms with data from a thread running in the background.
I used the SyncContext with the object Sync to lock my collection while i'm modifying it to prevent from the exception "InvalidOperationException: Collection Was Modified", but when I try to change my view to another view, with another data context, I get a "System.ArgumentNullException: 'Value cannot be null.". I find out that it's related to the SyncContext because when I remove it from the XAML, my application doesn't crash, but the problem with the collection was modified crash it.

I think the ViewModel is kinda faster (I don't know how to explain clearly) than the view then when I load another view, I generate another ViewModel and put it in the data context of the app, replacing the old ViewModel where I have my chart, and I think the old view still needs the SyncContext to do something. Well it's really weird and I can't solve my problem alone.

**Comment by pkindruk** [comment]
Try `Mode=OneWayToSource'.

If that doesn't help can you post full stacktrace of `ArgumentNullException`? And how your binding to `SyncContext` looks like?

**Comment by MathieuGrdr** [comment]
> Try `Mode=OneWayToSource`.
> 
> If that doesn't help can you post full stacktrace of `ArgumentNullException`? And how your binding to `SyncContext` looks like?


I tried the OneWayToSource, but it crash when i'm initializing the graphe :    

at MS.Internal.Data.PropertyPathWorker.CheckReadOnly(Object item, Object info)
   at MS.Internal.Data.PropertyPathWorker.ReplaceItem(Int32 k, Object newO, Object parent)
   at MS.Internal.Data.PropertyPathWorker.UpdateSourceValueState(Int32 k, ICollectionView collectionView, Object newValue, Boolean isASubPropertyChange)
   at MS.Internal.Data.ClrBindingWorker.AttachDataItem()
   at System.Windows.Data.BindingExpression.Activate(Object item)
   at System.Windows.Data.BindingExpression.AttachToContext(AttachAttempt attempt)
   at System.Windows.Data.BindingExpression.MS.Internal.Data.IDataBindEngineClient.AttachToContext(Boolean lastChance)
   at MS.Internal.Data.DataBindEngine.Task.Run(Boolean lastChance)
   at MS.Internal.Data.DataBindEngine.Run(Object arg)
   at MS.Internal.Data.DataBindEngine.OnLayoutUpdated(Object sender, EventArgs e)
   at System.Windows.ContextLayoutManager.fireLayoutUpdateEvent()
   at System.Windows.ContextLayoutManager.UpdateLayout()
   at System.Windows.ContextLayoutManager.UpdateLayoutCallback(Object arg)
   at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
   at System.Windows.Media.MediaContext.RenderMessageHandlerCore(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.RenderMessageHandler(Object resizedCompositionTarget)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(Object state)
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at MS.Internal.CulturePreservingExecutionContext.Run(CulturePreservingExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.Run()
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at System.Windows.Application.Run()
   at BoxAuto_LT_M.App.Main() in U:\R_I\RI-A210003_304-AVATAR_II-FLUX_5_LOGICIEL_D_AIDE_AU_CHOIX\DEV_C#\BA_4_M\BoxAuto_LT_M\BoxAuto_LT_M\obj\Debug\net6.0-windows\App.g.cs:line 53 


I'm binding my SyncContext like that : 
In my view
![image](https://user-images.githubusercontent.com/62133663/205584817-808b3bfc-263b-460b-9119-b8259d1309bf.png)

And I'm using it like that in my viewmodel
![image](https://user-images.githubusercontent.com/62133663/205587033-af931c2e-f571-4949-8170-7d199e47b972.png)


there is the full stacktrace of the ArgumentNullException

   at System.Threading.Monitor.ReliableEnter(Object obj, Boolean& lockTaken)
   at LiveChartsCore.Chart`1.<UpdateThrottlerUnlocked>b__151_1()
   at System.Windows.Threading.DispatcherOperation.InvokeDelegateCore()
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
--- End of stack trace from previous location ---
   at System.Windows.Threading.DispatcherOperation.Wait(TimeSpan timeout)
   at System.Windows.Threading.Dispatcher.InvokeImpl(DispatcherOperation operation, CancellationToken cancellationToken, TimeSpan timeout)
   at System.Windows.Threading.Dispatcher.Invoke(Action callback, DispatcherPriority priority, CancellationToken cancellationToken, TimeSpan timeout)
   at System.Windows.Threading.Dispatcher.Invoke(Action callback)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at LiveChartsCore.Kernel.ActionThrottler.<Call>d__8.MoveNext()
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(Object state)
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at MS.Internal.CulturePreservingExecutionContext.Run(CulturePreservingExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.Run()
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at System.Windows.Application.Run()
   at BoxAuto_LT_M.App.Main() in U:\R_I\RI-A210003_304-AVATAR_II-FLUX_5_LOGICIEL_D_AIDE_AU_CHOIX\DEV_C#\BA_4_M\BoxAuto_LT_M\BoxAuto_LT_M\obj\Debug\net6.0-windows\App.g.cs:line 53

**Comment by pkindruk** [comment]
`ArgumentNullException` looks like I expected. How do you replace your ViewModel with a new one? I can't understand why you get null anyway.

What you are doing is trying to rewrite Canvas sync object. However each Canvas already has a sync object. With `OneWayToSource` you can copying it to your ViewModel. As for your ViewModel it is missing setter `public object Sync { get; set; }` to work with this option.

**Comment by MathieuGrdr** [comment]
To be honest, I'm newbie to C# WPF, and I think that my MVVM is not clean.

Each time I change the view in my app, I change the datacontext of my main view : 
![image](https://user-images.githubusercontent.com/62133663/205623292-366ed495-c6a5-4f14-adb5-1fb6dfe7c32c.png)

Then if I change my view, I'm generating a new viewModel and change the data context with the new one.

I tried to add setter to my syncObject, but : 
![image](https://user-images.githubusercontent.com/62133663/205623708-3c85e805-0531-488e-89b0-1042bbc3324f.png)
   
at System.Threading.Monitor.ReliableEnter(Object obj, Boolean& lockTaken)
   at System.Threading.Monitor.Enter(Object obj, Boolean& lockTaken)
   at BoxAuto_LT_M.ViewModels.GraphesViewModel.set_ListSensorGr(List`1 value) in U:\R_I\RI-A210003_304-AVATAR_II-FLUX_5_LOGICIEL_D_AIDE_AU_CHOIX\DEV_C#\BA_4_M\BoxAuto_LT_M\BoxAuto_LT_M\ViewModels\GraphesViewModel.cs:line 125
   at BoxAuto_LT_M.ViewModels.ClassConn.BoucleduThread() in U:\R_I\RI-A210003_304-AVATAR_II-FLUX_5_LOGICIEL_D_AIDE_AU_CHOIX\DEV_C#\BA_4_M\BoxAuto_LT_M\BoxAuto_LT_M\ViewModels\ClassConn.cs:line 267
   at System.Threading.Thread.StartHelper.Callback(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)

**Comment by pkindruk** [comment]
`OneWayToSource` is a dead end. I got confused in init code for chart control.

I tried reproducing issue based on your description. But it works on my side. Is there any chance that you set your DataContext to null somewhere? Is there a minimal repro that you can share?

I will attach what I tried. [TestLvc785.zip](https://github.com/beto-rodriguez/LiveCharts2/files/10170322/TestLvc785.zip)

I could reproduce crash behavior only when I set DataContext to null and leave UI thread to let chart try to render something.
```cs
private void Btn1_Click(object sender, RoutedEventArgs args)
{
    // this will crash
    // DataContext = null;
    // await Task.Delay(100);
    DataContext = new SimpleViewModel();
}
```

**Comment by MathieuGrdr** [comment]
Hi, I checked the whole day yesterday, but I can't figure out, why it's isn't working on my project.

DataContext is never null, I did step by step but never see that.

I can't share you the repro of my project because there is a lot of dependencies.

**Comment by MathieuGrdr** [comment]
Even if I don't launch my thread to grab my datas, I'm just switching between views, and it crash

**Comment by MathieuGrdr** [comment]
I managed to isolate the problem, there is a repro : https://dropmefiles.com/EG27H (too big to be in the post)

You just have to launch the app, click on "Blue view" button, and then click on "Orange View" button.

I started from my entire project and then delete files one by one to see if there was a dependency problem with other library, but it don't seem so.

PS : rebuild solution if needed

**Comment by pkindruk** [comment]
Seems that issue is kind of fixed in `beta.603`. Null refence is still there, but now it is silently ignored. Before it was thrown all the way to `ThreadPool` which eventually killed the app.

P.S. For future if you want to share a repro it is better to do a Clean which clears `bin`, `obj` directories those are generated during build anyway. Also there is a hidden `.vs` directory which is not required for building. The required part was about 120 Kb

**Comment by MathieuGrdr** [comment]
Thank you, when will the update be available ?

**Comment by pkindruk** [comment]
Your case should work on `beta.603` already available. For new version you better tag repository owner

**Comment by beto-rodriguez** [comment]
Will be included in beta 700, should be available in NuGet in a few hours, this should be fixed thanks to #811, let us know if the issue continues in beta 700, I will close this for now.
